### ISSUE TITLE
Shouldn't deduce NotImplemented as a possible return type for arithmetic and comparison functions

### ISSUE THREAD

**Issue by MC-DeltaT** [microsoft/pyright#4473]
**Describe the bug**
When overloading an arithmetic or comparison operator and that can return `NotImplemented`, and you don't explicitly write a return type, Pyright seems to deduce that `NotImplemented` is a possible return type. This then causes type checking issues when using the result of the operator.
I believe this should not be the case, since the result will never actually be `NotImplemented` - Python will raise a `TypeError` instead.

**To Reproduce**
1. Overload an arithmetic or comparison operator.
2. Add a type hint for the operator RHS argument.
3. Don't add a type hint for the operator return type.
4. Conditionally return `NotImplemented` from the operator.
5. Invoke the operator with a valid argument type.
6. The type of the return value includes `NotImplemented`.

**Expected behavior**
The type of the return value should not include the possibility of `NotImplemented`, since the caller can never receive `NotImplemented` as a result.

**Screenshots or Code**
```python
class MyClass:
    def __add__(self, other: 'MyClass'):
        if isinstance(other, MyClass):
            return 42
        else:
            return NotImplemented


# Type of x is _NotImplementedType | Literal[42]
x = MyClass() + MyClass()
# Type of y is partially unknown
y = x / 10
```

The type of `x` should be exactly `Literal[42]` and the type of `y` should be exactly `float`.

![image](https://user-images.githubusercontent.com/25008419/212853718-c106f0c1-7dc4-499f-afd6-d1da7314c049.png)

**VS Code extension or command-line**
Running Pyright via the Pylance VS Code extension version v2023.1.20.

**Comment by erictraut** [comment]
Pyright is working as designed here. If you omit a function's return type annotation, pyright infers its return type based on the return and yield statements within the function implementation. Adding special cases for certain categories of methods is something we try to avoid because it leads to inconsistent and confusing behavior and is a maintenance burden. The recommended approach here is to provide an explicit return type annotation.

**Comment by MC-DeltaT** [comment]
Personally I think there is still value because of some scenarios where having an explicit return type is difficult. E.g. returning the same type as the object:

```python
class Foo:
    def __add__(self, other: 'Foo'):
        return type(self)(...)
```

Before Python 3.11 that is cumbersome to explicitly type hint. I'd rather the type checker figures it out itself.

In regards to special behaviour, doesn't pyright already have some special behaviour to handle reversed arithmetic methods, e.g. `__radd__`? (Which has a significantly complex set of rules behind it.)
Omitting `NotImplemented` from the set of return types seems like pretty straightforward behaviour; IMO the current behaviour is much more confusing given that the caller can never receive `NotImplemented` as a return value.

Perhaps I could convert this to a feature request instead?

**Comment by oscarbenjamin** [comment]
I think the problem is that pyright conflates the expression `a + b` with the function call `a.__add__(b)` but these two things are not the same. It should be modelled that `a + b` is equivalent to `add(a, b)` where the `add` function is defined as:
```python
def add(a, b):
    result = a.__add__(b)
    if not isinstance(result, NotImplementedType):
        return result
    result = b.__radd__(a)
    if not isinstance(result, NotImplementedType):
        return result
    raise TypeError
```
It is guaranteed by the runtime that the expression `a + b` can never evaluate to `NotImplemented`.

**Comment by oscarbenjamin** [comment]
If you implement the `add` function explicitly then pyright can infer the type:
```python
from __future__ import annotations

from types import NotImplementedType
from typing import overload, Protocol, TypeVar

T1 = TypeVar('T1', contravariant=True)
T2 = TypeVar('T2', covariant=True)

class HasAdd(Protocol[T1, T2]):
    def __add__(self, other: T1, /) -> T2 | NotImplementedType:
        ...

class HasRAdd(Protocol[T1, T2]):
    def __radd__(self, other: T1, /) -> T2 | NotImplementedType:
        ...

@overload
def add(a: HasAdd[T1, T2], b: T1) -> T2: ...
@overload
def add(a: T1, b: HasRAdd[T1, T2]) -> T2: ...

def add(a, b):
    result = a.__add__(b)
    if not isinstance(result, NotImplementedType):
        return result
    result = b.__radd__(a)
    if not isinstance(result, NotImplementedType):
        return result
    raise TypeError


class MyClass:
    def __add__(self, other, /):
        if isinstance(other, MyClass):
            return 42
        else:
            return NotImplemented


reveal_type(MyClass.__add__)
# (self: MyClass, other: Unknown, /) -> (_NotImplementedType | Literal[42])

a = MyClass()
b = MyClass()
reveal_type(add(a, b)) # int

reveal_type(add(1, 2)) # int
reveal_type(add(1.0, 2)) # float
reveal_type(add(1, 2.0)) # float
reveal_type(add(1j, 2.0)) # complex
reveal_type(add([2], [])) # list[int]
reveal_type(add("asd", "234")) # str
reveal_type(add([], ())) # error
reveal_type(add(1, ())) # error
```
